pkg/kv/kvserver/replica_proposal_quota.go

```diff
@@ -164,9 +164,12 @@ func (r *Replica) updateProposalQuotaRaftMuLocked(
 		) {
 			return
 		}
-
-		// Only consider followers that that have "healthy" RPC connections.
-		if err := r.store.cfg.NodeDialer.ConnHealth(rep.NodeID, r.connectionClass.get()); err != nil {
+		// At this point, we know that either we communicated with this replica
+		// recently, or we became the leader recently. The latter case is ambiguous
+		// w.r.t. the actual state of that replica, but it will be disambiguated
+		// soon. We do a second check using the NodeLiveness to increase the chance
+		// of disambiguating here.
+		if !r.store.cfg.NodeLiveness.IsAvailable(rep.NodeID) {
```

Comment:
Alternatives:

1. Only do this check if we recently became the leader, because otherwise `isFollowerActiveSince` does the job.
1. Only do the `NodeLiveness` check. Are there cases when it would not detect things that per-replica activity detects?
1. Leave the `ConnHealth`, but gate it by "I recently became the leader", similarly to (1). Are there cases when `ConnHealth` is more accurate than `NodeLiveness`?

Reply:
I think the case we're interested in here is where the follower replica is active, and the node is up and running, but the node is struggling and is failing its liveness heartbeats. This is unrelated to whether we just became the leader or not -- even if we've been the leader for a long time, and `isFollowerActiveSince` returns `true`, if the node is struggling then we don't want it to hold up the range.

One typical example of this is a stalled disk, since we do a synchronous disk write during node heartbeats specifically to check this (stalled disks are a common failure mode with networked disks used in cloud VMs):

https://github.com/cockroachdb/cockroach/blob/eeb72360fe8e5bdc9ed123612c4e71b975b52124/pkg/kv/kvserver/liveness/liveness.go#L1286-L1288

So we should have both checks:

* `isFollowerActiveSince` will check that the Raft replica is live.
* `IsAvailable` will check that the node is live and reasonably healthy.

This is related to a broader discussion on whether we even want the quota pool to ever hold up the range at all (#75978). Most of the time, we really don't. But for now, this is one step in the right direction.

I'll also note that `IsAvailable()` will return false if the node is not in the local liveness cache (i.e. we don't know about it). This seems fine, since that will default to not holding up the range.

Reply:
Is this disk write on the requested node, or on some other range/nodes where the liveness records are stored?

Reply:
This happens on every node every time they send a heartbeat.

For reference, liveness records themselves are stored as key/value pairs in a range, just like any other data in the system. So there will be one leaseholder somewhere in the cluster who's responsible for them.

Reply:
Ah, got it.

Reply:
What kinds of events refresh `isFollowerActiveSince` timestamps? Are these successful Raft RPCs, or any interactions?
Any example when `isFollowerActiveSince` is true, but the node is not live?

Reply:
> What kinds of events refresh `isFollowerActiveSince` timestamps? Are these successful Raft RPCs, or any interactions?

Any Raft message received from the follower:

https://github.com/cockroachdb/cockroach/blob/d10bde7a099716fcc4c3270c67996899fb121a01/pkg/kv/kvserver/replica_raft.go#L517-L518

>Any example when `isFollowerActiveSince` is true, but the node is not live?

A simple but contrived example might be a node with a partial network partition: it's not able to communicate with the liveness leaseholder (so it can't heartbeat), but it is able to communicate with other nodes. This isn't all that relevant, because in this case the node itself isn't necessarily struggling as such, but it's a fairly clear example.

A more relevant example might be a very overloaded node. Maybe it's taking more that 4.5 seconds (the heartbeat timeout) to fsync the disk, so node heartbeats aren't being sent. Or maybe the heartbeat goroutine isn't getting scheduled or the network is backlogged. However, it is able to occasionally/eventually append Raft proposals and respond to Raft leaders. In these cases, replicas on that node would be very slow to append Raft proposals, and we would rather sacrifice the followers on that node than slow down the entire range.